### ISSUE TITLE
Documentation For 'NanAsyncProgressWorker' and "NanAsyncBufferWorker"

### ISSUE THREAD

**Issue by dynamite-ready** [rvagg/nan#271]
As I understand, NanAsyncProgressWorker is a new NAN feature, I also suspect that NanAsyncBufferWorker is new, but can't be sure. The tests contain an implementation for both classes, but while I have a rough idea of what NanAsyncProgressWorker does from it's name, I find it hard to work out what NanAsyncBuffer is for. 

What I would like to do, is:
 
- Call a Javascript function with a callback 
- Run an indefinite loop in a C++ async 
- Have that loop change a handful of static values
- Use either a NanAsyncProgressWorker or NanAsyncBufferWorker to retrieve those static values each time they change 
- Wrap those values up in an object 
- Return the object as a callback parameter.

Is this possible?

**Comment by kkoopa** [comment]
You are right in that `NanAsyncProgressWorker` is a new feature. However, `NanAsyncBufferWorker` is not, it is just a realization of `NanAsyncWorker` that is part of a test.

While it might somehow be possible to do what you want with `NanAsyncProgressWorker`, that is not its intended use. It is meant to send progress notifications about long running tasks so you can update a progress bar.

Why not just give in a callback that you can invoke whenever those magic values are supposed to change?

**Comment by dynamite-ready** [comment]
Not sure if this is the best place to discuss it, but it's a domain with very little coverage on Stackoverflow... Right now, I: 

- Feed in a Javascript callback to a 'run' method.
- I kick off an NanAsyncWorker with a dummy callback (i.e, NaNAsyncWorker).
- The NanAsyncWorker kicks of an infinite while loop in a persistent worker, which updates a struct.
- In the main thread, I run another infinite 'event' loop, which contains a conditional that will run the Javascript callback function when key struct values change, and return an object that reflects the current state of the static values inside the struct I'm interested in.

It works, but there's a definite memory leak. 
I think the infinite loop in the main thread is preventing the V8 garbage collector from doing what it usually does, so copies of the Javascript object that I create for the callback are being held in memory...

I wanted to make the JavaScript callback from the code kicked off by the NanAsyncWorker, but the manual suggests that any code inside a worker should not call or create any V8 stuff... Am I right?

Some advice will be great, because I'm sure having the two loops is a bad idea. Hence why I'm considering the use of an NanAsyncProgressWorker.

**Comment by dynamite-ready** [comment]
Ok... So I've optimized my application, based on this conversation, and now have a nice clean module that only runs a single long running loop in an AsyncWorker, 

interestingly though, when I run the module, I create a new instance of the object that the module produces (it's a .js constructor function from), run it for a while (collecting the yield in a setInterval method), delete it, and then open a new instance. I do this periodically.

After the fifth cycle though, the worker slows down to a shocking rate... Almost as if the garbage collector had just woken up. I don't think it's the loop, as I've noticed that the loop will block the entire application if it's not explicitly stopped (or crashes).

Is this not unheard of? I've noticed a similar bug when running Node Serialport on Windows.
Especially when repeatedly connecting to the same serial port and device.

For example, one question I have is how in this small example;

    // Spawn a worker thread...
    NanCallback* dummyCallback = new NanCallback(args[0].As<Function>());
    NanAsyncWorker* workerThread = new DeviceDataThread(dummyCallback, device);
    NanAsyncQueueWorker(workerThread);

Will the NanAsyncWorker take care of deleting the "workerThread" internally, or is that my responsibility?

**Comment by kkoopa** [comment]
`NanAsyncQueueWorker` destroys `workerThread` upon completion, taking `dummyCallback` with it.

**Comment by dynamite-ready** [comment]
Thank you for the clarification. I'm kinda' keen on examining this stuff fully, as it I've turned to the idea of writing a native module, because of a fault I found in Node Serialport. 

So in observing this consistent slow down both in my application, and in Node Serialport (which also uses NAN), I wonder if both issues have the same root cause...

I can't, of course, rule out my own stupidity and inexperience, but what I've written, while a little uncommon, is not overly complicated (it essentially reads a stream from a serial port device... Another common trait between my app, and Node Serialport). It should also be noted that I'm running the app through Node Webkit (v0.7.5) (as I also did with Node Serialport, when I encounter the same problems with slowdown).

Although it's unlikely, the only article that matches my current suspicion is this:
https://www.joyent.com/blog/walmart-node-js-memory-leak

The article alludes to a scale much larger than what I'm working with, but as I'm polling a serial port device (@9600Bd), I would very likely parse enough data to encounter something similar, no?

And failing that line of investigation, what would be a typical cause of the type of slowdown I'd described?

Thank you for the help up to this point, btw.
